containers/wordpress/run.sh

```diff
@@ -0,0 +1,184 @@
+        tmp_name=cpusets_socket${socket_id}
+        all_cpus=($(echo ${!tmp_name} | tr ' ' '\n' | cat -n | sort -uk2 | sort -n | cut -f2- | tr '\n' ' '))
+        if [[ ${socket_id} -eq $((${sockets_num} - 1)) ]]; then
+            count_per_socket=$(($count_per_socket + $COUNT % $sockets_num))
+        fi
+        cpu_per_instance=$((${#all_cpus[@]} / $count_per_socket))
+        start=0
+        for ((i = 1; i <= $count_per_socket; i++)); do
+            if [[ ${i} -eq $count_per_socket ]]; then
+                array=("${all_cpus[@]:${start}}")
+            else
+                array=("${all_cpus[@]:${start}:${cpu_per_instance}}")
+            fi
+            start=$((start + cpu_per_instance))
+            cpuset_cpus_s=$(printf ",%s" "${array[@]}")
+            cpuset_cpus_s=${cpuset_cpus_s:1}
+
+            docker_run_s="--cpuset-cpus="${cpuset_cpus_s}" --cpuset-mems="${socket_id}
+            cpu_pinning_s+=("$docker_run_s")
+        done
+    done
+}
+
+# Calculate the worker number of each instance. Total workers number equals to 1.5X core num on this system.
+core_count=$(nproc)
+worker_num=$(echo "${core_count} * 1.5 / ${COUNT} " | bc -l)
+worker_num=${worker_num%.*}
+
+# Create temp directory for container output
+tmp_dir=$(mktemp -d -t run-XXXXXXXXXX)
+echo "-------------------------------------------------------------"
+echo "Creating temporary directory ${tmp_dir} for logfile."
+echo ""
+
+# Launch containers
+all_containers=()
+if [[ ${NUMA_PINNING} -eq 0 ]]; then
+    echo "-------------------------------------------------------------"
+    echo "Running ${COUNT} ${IMAGE_NAME} instance(s)."
+    echo ""
+    for ((i = 0; i < ${COUNT}; i++)); do
+        container_id=$(sudo docker run -d --rm -ti --privileged ${IMAGE_NAME} bash -c "./quickrun.sh \"--php-fcgi-children ${worker_num}\"")
+        all_containers[$i]=$(sudo docker ps -q -f id=${container_id})
+    done
+else
+    echo "-------------------------------------------------------------"
+    echo "Running ${COUNT} ${IMAGE_NAME} instance(s) with NUMA pinning."
+    echo ""
+    get_cpu_pinning
+    for ((i = 0; i < ${COUNT}; i++)); do
+        container_id=$(sudo docker run -d --rm -ti --privileged ${cpu_pinning_s[i]} ${IMAGE_NAME} bash -c "./quickrun.sh \"--php-fcgi-children ${worker_num}\"")
+        all_containers[$i]=$(sudo docker ps -q -f id=${container_id})
+    done
+fi
+
+# Redirect containers output to logfile in temp directory
+for container_id in ${all_containers[@]}; do
+    sudo docker logs -f ${container_id} >${tmp_dir}/${container_id}.log &
+done
+
+# Wait all container to be completed
+while true; do
+    completed=0
+    for container_id in ${all_containers[@]}; do
+        if [ "$(sudo docker ps -q -f id=${container_id})" ]; then
+            completed=1
+            echo "Container ${container_id} = running"
+        else
+            echo "Container ${container_id} = completed"
+        fi
+    done
```

Comment:
Does this cause overhead with this loop spinning? Is it worth having a pause/sleep here?

Reply:
Resolve now at line 169